### ISSUE TITLE
REQ Secure user isolation thru namespaces and cgroups. Alternative to CageFS.

### ISSUE THREAD

**Issue by chris001** [virtualmin/virtualmin-gpl#432]
Ordinary "server and subserver owners", and their users, should only see their own files/dirs, plus the most minimal possible essential system files/dirs (`/var/log/`...), symlinked by default thru their home dir, same as `jailkit` and `cagefs`.

Users should not see other usernames (`passwd`, `shadow`).

How interesting would it be for Virtualmin to consider developing a secure user isolation package, combining open source software `namespaces` + `cgroups` = equivalent to almalinux's original sponsor's (cpanel/cloudlinux) $10/mo subscription package `cagefs`.  For Pro only, a great reason to subscribe to Pro.  Or a subscription for Pro and GPL.

Ability to block full machine server data theft hacks that enter thru one website with a flawed plugin.  Reasonable to expect it would be about as popular as `cagefs` and grow the base of Pro subscribers.

**Comment by jcameron** [comment]
We do have some of this functionality already in Virtualmin's jailkit feature - users in a jail should only be able to see their own domain's accounts in /etc/passwd .

**Comment by chris001** [comment]
`jailkit` is a start.

How about all files on the system, making all files/dirs, which they don't have "file permissions" to (i.e. match user, member of group, or the file or dir has have world permission), seem invisible, called "`namespaces`"?

Seems `jailkit` and `namespaces` should be enabled by default in Virtualmin, at least for non admins, e.g. users, subservers, servers.  Only system admins should be able to see every file/dir, even those they don't have perms on.

**Comment by jcameron** [comment]
I'm not actually familiar with `namespaces` - I'd have to look into this and see how we could incorporate it.

What benefits do they give over jails though? A user in a jail as setup by Virtualmin cannot even see other domain's home dirs.

**Comment by chris001** [comment]
[Here's a nice deep dive on these user isolation technologies.](https://www.youtube.com/watch?v=pTlBxn-QAnw)

**Comment by jcameron** [comment]
My gut feel is that these kinds of isolation methods are a half-measure that doesn't protect against local privilege exploits, and aren't worth the effort to implement. The better alternative is full virtualization which has much harder isolation, or just relying on regular Linux permissions which Virtualmin already sets up (ie. domain owners cannot see each others directories).

**Comment by chris001** [comment]
True, and maybe these need to develop more, however, full virtualization's all or nothing approach, while it offers total separation, has heavy exclusive system resource requirements which many don't have available.   Many want a shared server with the high peak performance of jails or namespace/cgroup isolation.  It should be possible with minimal lines of code to make each virtual server and sub-server, live inside its own jail or namespace, or lightweight container with its own IPv6 address, which are free and abundant.  Regular Linux permissions are great, however they let users see file/dir names which they have no access rights to, and namespaces keeps those file/dir names invisible, reducing the attack surface and preventing information leakage.

**Comment by chris001** [comment]
[What are Namespace and Cgroups and how do they work](https://www.nginx.com/blog/what-are-namespaces-cgroups-how-do-they-work/)

**Comment by jcameron** [comment]
Cgroups is actually one feature that's on our TODO list already - I feel like that's more useful in practice than filesystem isolation, as it protects CPU and RAM resources in a shared hosting environment.

**Comment by chris001** [comment]
Yes!  It could be optional the isolation.  With probably 10 lines of code it should be possible to try out what it would be like to put a Virtualmin virtual server into a well supported minimal lightweight container system, such as docker, and see if it breaks anything.

**Comment by jcameron** [comment]
Sadly I think it would be way more than 10 lines of code in Virtualmin to implement any of these features. But cgroups shouldn't be *too* hard.

**Comment by chris001** [comment]
If using a high level engine such as docker to containerize a virtual server, then it should need a limited number of commands to get it up and running.
https://github.com/rafi-java/Documentation/wiki/Linux-Containers#linux-containers

https://github.com/tanalam2411/docker/wiki/Linux-Namespaces

**Comment by chris001** [comment]
What if for every user who logs in via `PAM`, such as `sshd`, `usermin`, `proftpd`, etc., the user, by default, will enter their `jail`, or a minimal docker container, based in their `home` dir?
[Users login thru `PAM` into their own lightweight isolated docker container](https://github.com/flant/pam_docker).

**Comment by jcameron** [comment]
These are all good ideas, however they all have implementation cost and add complexity ... so they'd likely be in a later version.

**Comment by chris001** [comment]
The Virtualmin Jailkit package! [2016-2017 Virtualmin forum discussion of @swelljoe new CentOS 7 jailkit package vs modern Linux namespaces docker isolation.](https://forum.virtualmin.com/t/testers-wanted-for-jailkit-package/54999/4)

**Comment by chris001** [comment]
[Getting started with lightweight LXD containers full guide with step by step command lines!](https://www.youtube.com/watch?v=aIwgPKkVj8s) Fully isolated linux containers, very low overhead (benchmark: 600 lxd containers vs. 30 kvm full virtualized fit on the same 16GB 4 core intel server hardware.)!

**Comment by chris001** [comment]
[This also does the namespace and cgroups isolation, tested, supported, etc.](https://github.com/netblue30/firejail)

**Comment by jcameron** [comment]
Even if it's easy to set containers up manually, the tricky part is writing the code to automate it under all possible circumstances..

**Comment by chris001** [comment]
> tricky part is writing the code to automate it under all possible circumstances.

One could look at how `CageFS` automatically catches the user on login and puts them into a "cage" (container) under all possible circumstances.

Probably, every user login to a system service that browses and reads/writes the user's home filesystem: `sshd`, `ProFTPd`, maybe others,  (not the heavy global services: SMTP, IMAP, DNS, antivirus, antispam, etc) would go thru PAM to login, and use a lib, [`pam_docker`](https://github.com/flant/pam_docker))  or `pam_lxd`, etc., to intercept the login, and give the user a session inside their own lightweight container, not in the top level Linux OS environment, unless the admin disables the isolated containerized user logins.

**Comment by chris001** [comment]
[LxdMosaic, Web interface to manage multiple instances of LXD linux containers.](https://github.com/turtle0x1/LxdMosaic)

**Comment by chris001** [comment]
How to install LxdMosaic in its own LXD container!
On Ubuntu/debian based linux (steps also available for CentOS/rpm linux):

```
sudo apt install snapd
sudo snap install lxd

# Do the network bridge to the container.
sudo lxd initip route show # find your untagged vlan.
lxc profile create untagged
lxc profile device add untagged eth0 nic   # Use your actual network info. 
nictype=macvlan1 parent=qvs1  # use your machine's nic type and parent net name.

# Create container for LxdMosaic. Uses cgroups.
lxd launch ubuntu:lxdMosaic --profile default --profile untagged -c security.privileged=true -c limits.memory=1024MB -c limits.cpu.allowance=20% boot.autostart=true

# Enter LxdMosaic container and install LxdMosaic
lxc exec lxdMosaic bash
sudo apt install snapd
sudo snap install lxdmosaic
exit  # exit from LXD container back to host.

# Make LXD available over IP4 and IP6, and over all interfaces.
lxc config set core.https_address [::]
lxc config set core.trust_password some-password
```

In Browser go to to the container IP address, located at the end of the `LxdMosaic` installation log.

**Comment by chris001** [comment]
[LXD supports not only lightweight linux containers, but also full virtualization.. virtual machines!](https://linuxcontainers.org/lxd/getting-started-cli/#virtual-machines) @jcameron  Great for users with tons of system resources, and wants the all or nothing approach to user isolation.

**Comment by pixel-paul** [comment]
Doesn't Cloudmin already support LXC?

**Comment by chris001** [comment]
Yes, however, this is about adding isolating Virtualmin users to prevent them from seeing other users home dirs or processes or off limits system files.  When a user logs in, it would log them into their own lightweight container, in their home dir.

**Comment by pixel-paul** [comment]
Interesting, so each user has a FS container.

**Comment by chris001** [comment]
Plus, only safe binaries would be available to the user.
The user can't see any other users and would have no way to detect the presence of other users and their usernames on the server.
The user can't see server configuration files, such as Apache config files.
The user has a limited view of `/proc` filesystem and will not be able to see other users processes.
User’s environment is fully functional, the user will not feel restricted in any way, no modifications to user apps or scripts are needed.

**Comment by jcameron** [comment]
If an attacker compromises wordpress, he will only be able to run code as that domain's user. By default, permissions on directories under /home are setup to prevent the compromised user from accessing other directories - so he's be able to see that users exist, but not view or change their files.

That said, unfortunately most attackers just want to send spam or DDOS attacks, so even a perfect restricted environment wouldn't do much to prevent that :-(

**Comment by chris001** [comment]
Another popular hack - installing a crypto miner service, mining results to the hacker's account.
Emailed you about another category of attack, best not mentioned too much, before it's no longer effective.

**Comment by jcameron** [comment]
Yeah, crypto miners are a problem. That sounds like something cgroups can fix though, along with CPU wastage in general..

**Comment by chris001** [comment]
Ah-hah... there is an [open source LXD-dashboard container control panel](https://github.com/lxdware/lxd-dashboard) which code could be borrowed from under free open source license!

**Comment by chris001** [comment]
A [2019 paper](https://www.researchgate.net/publication/332927363_The_Concept_of_LXD-based_Web_Hosting_Panel) describes replacing non-isolated shared web hosting, with LXD container based shared hosting, for safe secure isolation.

**Comment by chris001** [comment]
[A set of lists of whitelist paths from the host that each user container should be given access to by default](https://github.com/im-richard/cagefs-cfg/tree/main/etc/cagefs/conf.d) so that user’s environment in the container be fully functional, the user would not feel restricted in any way, no modifications to user apps or scripts would be needed.

**Comment by chris001** [comment]
[BubbleWrap](https://github.com/containers/bubblewrap) and [Sandstorm.io](https://sandstorm.io/) also do user isolation.

**Comment by chris001** [comment]
[This issue, related to preventing BIND9 from listening on 3 particular local IP address ranges used by `lxd` which runs a local `DNSmasq` service that listens for `DNS` queries from its containers](https://github.com/lxc/lxd/issues/10874), is important to account for when expecting `lxd` containers networking to work without issue, on a Virtualmin shared hosting server.

**Comment by iliajie** [comment]
> [This issue, related to preventing BIND9 from listening on 3 particular local IP address ranges used by `lxd` which runs a local `DNSmasq` service that listens for `DNS` queries from its containers](https://github.com/lxc/lxd/issues/10874), is important to account for when expecting `lxd` containers networking to work without issue, on a Virtualmin shared hosting server.
>> As an experiment, same as with the IPv4 range `10.x.x.x`, I tell `BIND9` to NOT listen on port 53 (DNS) for those local IPv6 address range patterns that `lxdbr0` has.

What about just excluding the whole interface range and listening on `any`. Would that work for your case?

```js
listen-on {
    !10.192.93.1/24;
    any;
};
```

**Comment by chris001** [comment]
I had to also exclude 2 IPv6 ranges from BIND to listen on DNS port 53, which are the "link local" and "dynamic self config (?)" ranges, because the `lxd` bridge device `lxdbr0` needs to listen on these IP addresses, on DNS port 53:
```
tls local-tls {
   key-file  "/etc/letsencrypt/live/my.domain.tld/privkey.pem";
   cert-file "/etc/letsencrypt/live/my.domain.tld/fullchain.pem";
};

options {
	directory "/var/cache/bind";

	dnssec-validation auto;

	listen-on { !10.0.0.0/8; my.public.ipv4.addr; localhost; };
	listen-on tls local-tls { !10.0.0.0/8; my.public.ipv4.addr; localhost;  };
      listen-on-v6 { !fd00::/8; !fe80::/16; my:public:ipv6:addr; localhost; };
	listen-on-v6 tls local-tls { my.public.ipv6.addr; localhost; };
};
```

**Comment by marclaporte** [comment]
Related: I see this in Virtualmin Server templates: "Chroot jail new domain Unix users"
![2022-09-17_220111](https://user-images.githubusercontent.com/1004261/190882426-d8c676fb-560b-41c2-9720-62d066ded62e.png)

**Comment by chris001** [comment]
Thanks for sharing @marclaporte `chroot` it's better for users to be isolated than not!

More info on [the problem with chroot alone](https://btholt.github.io/complete-intro-to-containers/namespaces#the-problem-with-chroot-alone) this is why `namespaces` takes `chroot` to the next level by providing also `process ID` isolation between users, and `cgroups` limit excessive RAM and CPU usage!

**Comment by iliajie** [comment]
Virtualmin needs to look into the future, meaning that for the future development I would rather focus on bringing containers support, as our time and resources aren't unlimited.

For the purposes we currently have standard Unix user permission isolation and _chroot_ is more than enough. Anyone needing more isolation and resource control should use virtualization. There is no reason to make things more complex than they already are.

**Comment by chris001** [comment]
This would be for consideration for a future version of Virtualmin, when it would be just a few simple commands for Virtualmin to enable a PAM plugin which creates or reuses one `LXD` lightweight container per user at SSH/SFTP login, with limited CPU and RAM by `cgroups` through command line switches.  Equivalent to Cpanel CageFS, goals are resource control, also most important to prevent common priv esc attacks at host level from achieving a data breach of all data on the server.

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run.

⚠️ If the issue is resolved, please take a moment to close it to help keep the project organized.
